akka-projection-slick/src/main/scala/akka/projection/slick/internal/SlickOffsetStore.scala

```diff
@@ -27,22 +27,37 @@ import slick.jdbc.JdbcProfile
   import OffsetSerialization.toStorageRepresentation
   import profile.api._
 
-  def this(db: P#Backend#Database, profile: P) = this(db, profile, Clock.systemUTC())
+  def this(db: P#Backend#Database, profile: P) =
+    this(db, profile, Clock.systemUTC())
 
   def readOffset[Offset](projectionId: ProjectionId)(implicit ec: ExecutionContext): Future[Option[Offset]] = {
     val action =
-      offsetTable.filter(_.projectionId === projectionId.id).result.headOption.map { maybeRow =>
-        maybeRow.map(row => fromStorageRepresentation[Offset](row.offsetStr, row.manifest))
+      offsetTable.filter(_.projectionId === projectionId.id).result.map { maybeRow =>
+        maybeRow.map(row => (row.offsetStr, row.manifest))
       }
 
-    db.run(action)
+    val results = db.run(action)
+
+    results.map { offsetRows =>
+      if (offsetRows.isEmpty) None
+      else Some(fromStorageRepresentation[Offset](offsetRows))
+    }
   }
 
+  private def newRow(projectionId: ProjectionId, offsetStr: String, manifest: String, now: Instant): DBIO[_] =
+    offsetTable.insertOrUpdate(OffsetRow(projectionId.id, offsetStr, manifest, now))
+
   def saveOffset[Offset](projectionId: ProjectionId, offset: Offset)(
       implicit ec: ExecutionContext): slick.dbio.DBIO[Done] = {
-    val (offsetStr, manifest) = toStorageRepresentation(offset)
-    val now = Instant.now(clock)
-    offsetTable.insertOrUpdate(OffsetRow(projectionId.id, offsetStr, manifest, now)).map(_ => Done)
+    val now: Instant = Instant.now(clock)
+    // TODO: maybe there's a more "slick" way to accumulate DBIOs like this?
+    toStorageRepresentation(offset)
+      .foldLeft(None.asInstanceOf[Option[DBIO[_]]]) {
+        case (None, (offset, manifest))      => Some(newRow(projectionId, offset, manifest, now))
+        case (Some(acc), (offset, manifest)) => Some(acc.andThen(newRow(projectionId, offset, manifest, now)))
+      }
+      .get
+      .map(_ => Done)
```

Comment:
We need to add a `.transactionally` here. 

And we need to confirm that transactions are re-entrant in Slick. I'm quite sure they are. 

Inside the projection we do:

```scala
// transactional over both
userDBIO.flatMap(_ => offsetDBIO).transactionally

// transactional for use only
userDBIO.transactionally.flatMap(_ => offsetDBIO)
```

With a Seq of DBIOs, this will become

```scala
// transactional over both
userDBIO.flatMap(_ => offsetDBIO.transactionally).transactionally

// needs 2 independent transactions
userDBIO.transactionally.flatMap(_ => offsetDBIO.transactionally)
```

Reply:
Ok.

TODO

Reply:
I've made an update to save the offset rows with `DBIO.sequence`.  IIUC the `DBIO` are enlisted in a transaction in `SlickProjectionImpl.processEnvelopeAndStoreOffsetInSameTransaction` and no further action is required in `saveOffsets`.  Please confirm.

Reply:
That's fine, thanks